### DISCUSSION TITLE
How to add WMS to ui.leaflet Map?

### DISCUSSION THREAD

**Discussion by LGV-BIM-Leitstelle** [zauberzeug/nicegui#3682] [Q&A] [answered]
I have a link to a WMS service: https://geodienste.hamburg.de/HH_WMS_DOP_belaubt?SERVICE=WMS&REQUEST=GetCapabilities. 
I would like to add the "stadtplan" layer in "EPSG:4326" instead of OpenStreetMap, but I couldn't find a way to do it. Is it possible to use L.tileLayer.wms for this purpose? I tried it but it didn't work. 
Alternatively, is there another way to add a WMS layer to ui.leaflet map?

**Comment by falkoschindler** [comment, marked as answer]
Hi @LGV-BIM-Leitstelle,

It looks like Leaflet supports WMS layers. But somehow its API isn't very straight-forward. While other layers can be added generically via [`L.tileLayer(...)`](https://leafletjs.com/reference.html#tilelayer) (supported by `ui.leaflet` via `tile_layer(...)`, WMS layers are created like [`L.tileLayer.wms(...)`](https://leafletjs.com/examples/wms/wms.html). So I don't see a simple way to use NiceGUI's existing API for WMS.

But you can try running some custom JavaScript to add the WMS layer like this:
```py
@ui.page('/')
async def page():
    m = ui.leaflet(center=(53.551086, 9.993682))
    m.clear_layers()
    await m.initialized()
    ui.run_javascript(f'''
        const wmsLayer = L.tileLayer.wms('https://geodienste.hamburg.de/HH_WMS_DOP_belaubt', {{
            layers: 'DOP_belaubt',
            format: 'image/png',
            transparent: true,
            version: '1.3.0',
            attribution: "© Freie und Hansestadt Hamburg, Landesbetrieb Geoinformation und Vermessung"
        }});
        getElement({m.id}).map.addLayer(wmsLayer);
    ''')
```

**Reply by LGV-BIM-Leitstelle** [reply]
This worked perfectly, thank you!
